Assets/SEE/GraphProviders/VCSGraphProvider.cs

```diff
@@ -0,0 +1,421 @@
+        protected void CheckArguments(AbstractSEECity city)
+        {
+            if (string.IsNullOrEmpty(RepositoryPath.Path))
+            {
+                throw new ArgumentException("Empty repository path.\n");
+            }
+            if (!Directory.Exists(RepositoryPath.Path))
+            {
+                throw new ArgumentException($"Directory {RepositoryPath.Path} does not exist.\n");
+            }
+            if (string.IsNullOrEmpty(CommitID))
+            {
+                throw new ArgumentException("Empty newCommitID.\n");
+            }
+            if (city == null)
+            {
+                throw new ArgumentException("The given city is null.\n");
+            }
+        }
+
+        /// <summary>
+        /// Builds the VCS graph with specific metrics.
+        /// </summary>
+        /// <param name="pathGlobbing">The paths which get included/excluded.</param>
+        /// <param name="repositoryPath">The path to the repository.</param>
+        /// <param name="commitID">The commitID where the files exist.</param>
+        /// <returns>the graph.</returns>
+        static Graph GetVCSGraph(Dictionary<string, bool> pathGlobbing, string repositoryPath, string commitID)
+        {
+            string[] pathSegments = repositoryPath.Split(Path.DirectorySeparatorChar);
+
+            Graph graph = new(repositoryPath, pathSegments[^1]);
+
+            // The main directory.
+            GraphUtils.NewNode(graph, pathSegments[^1], "directory", pathSegments[^1]);
+
+            IEnumerable<string> includedFiles = pathGlobbing
+                .Where(path => path.Value == true)
+                .Select(path => path.Key);
+
+            IEnumerable<string> excludedFiles = pathGlobbing
+                .Where(path => path.Value == false)
+                .Select(path => path.Key);
+
+            using (Repository repo = new(repositoryPath))
+            {
+                LibGit2Sharp.Tree tree = repo.Lookup<Commit>(commitID).Tree;
+                // Get all files using "git ls-tree -r <CommitID> --name-only".
+                IEnumerable<string> files;
+                if (includedFiles.Any() && !string.IsNullOrEmpty(includedFiles.First()))
+                {
+                    files = ListTree(tree).Where(path => includedFiles.Contains(Path.GetExtension(path)));
+                }
+                else if (excludedFiles.Any())
+                {
+                    files = ListTree(tree).Where(path => !excludedFiles.Contains(Path.GetExtension(path)));
+                }
+                else
+                {
+                    files = ListTree(tree);
+                }
+
+                // Build the graph structure.
+                foreach (string filePath in files.Where(path => !string.IsNullOrEmpty(path)))
+                {
+                    string[] filePathSegments = filePath.Split(Path.AltDirectorySeparatorChar);
```

Comment:
Why is the `AltDirectorySeparatorChar` used here instead of the normal one?

Reply:
The Paths we get from git are all with forward slashes, but DirectorySeparatorChar is backslash on Windows and forward slash on Unix-based systems. AltDirectorySeperatorChar is forward slash on Windows and Unix-based systems. So when we try to split the path to get the directories with DirectorySeparatorChar, all we get is the full path.
LibGit2Sharp paths output:
"Assets/Editor/WorldTransform.cs"
filePathSegments output with Path.AltDirectorySeparatorChar:
["Assets","Editor","WorldTransform.cs"]
filePathSegments output with Path.DirectorySeparatorChar on Windows:
["Assets/Editor/WorldTransform.cs"]